modules/launcher/src/main/java/org/mule/runtime/module/launcher/ApplicationDescriptorFactory.java

```diff
@@ -108,24 +109,77 @@ public ApplicationDescriptor create(File artifactFolder) throws ArtifactDescript
     {
         final List<ApplicationPluginDescriptor> result = new LinkedList<>(applicationPluginRepository.getContainerApplicationPluginDescriptors());
         result.addAll(plugins);
+
+        // Sorts plugins by name to ensure consistent deployment
+        result.sort(new Comparator<ApplicationPluginDescriptor>()
+        {
+            @Override
+            public int compare(ApplicationPluginDescriptor descriptor1, ApplicationPluginDescriptor descriptor2)
+            {
+                return descriptor1.getName().compareTo(descriptor2.getName());
+            }
+        });
+
         return result;
     }
 
     private void verifyPluginExportedPackages(List<ApplicationPluginDescriptor> plugins)
     {
-        final HashMap<String, String> exportedPackages = new HashMap<>();
+        final Map<String, List<String>> exportedPackages = new HashMap<>();
 
+        boolean error = false;
         for (ApplicationPluginDescriptor plugin : plugins)
         {
-            for (String exportedPackage : plugin.getClassLoaderFilter().getExportedClassPackages())
+            for (String packageName : plugin.getClassLoaderFilter().getExportedClassPackages())
+            {
+                List<String> exportedOn = exportedPackages.get(packageName);
+
+                if (exportedOn == null)
+                {
+                    exportedOn = new LinkedList<>();
+                    exportedPackages.put(packageName, exportedOn);
+                }
+                else
+                {
+                    error = true;
+                }
+                exportedOn.add(plugin.getName());
+            }
+        }
+
+        if (error)
+        {
+            throw new DuplicateExportedPackageException(buildPackageDuplicationErrorMessage(exportedPackages));
+        }
+    }
+
+    private String buildPackageDuplicationErrorMessage(Map<String, List<String>> exportedPackages)
```

Comment:
Move this to DuplicateExportedPackageException. Ideally the exception could have the exportedPackages map and build the message on demand, so that while debugging/inspecting it can be accessed and not just the string message.